src/rgb/contract/assignments.rs

```diff
@@ -708,3 +708,813 @@ mod strict_encoding {
+
+        // Extract balanced confidential output amounts
+        let outputs: Vec<Commitment> = balanced
+            .all_state_pedersen()
+            .iter()
+            .map(|confidential| confidential.commitment)
+            .collect();
+
+        // Create confidential input amounts
+        let inputs: Vec<Commitment> = input_revealed
+            .iter()
+            .map(|revealed| revealed.conceal().commitment)
+            .collect();
+
+        // Check sum verification and return the result
+        amount::Confidential::verify_commit_sum(inputs, outputs)
+    }
+
+    #[test]
+    fn test_zero_balance() {
+        let mut rng = thread_rng();
+
+        // test equal inputs and outputs
+        // 4 values equally distributed among ours and theirs allocations
+        // * One zero value case
+        // * All zero value case
+        // * One u64::MAX value case
+        // * All u64::MAX value case
+        // * One zero value + One u64::MAX value case
+        let amounts = vec![
+            [1u64, 2u64, 3u64, u64::MAX],
+            [10u64, 20u64, 30u64, 40u64],
+            [0u64, 197642u64, u64::MAX, 476543u64],
+            [0u64, 0u64, 0u64, 0u64],
+            [u64::MAX, u64::MAX, u64::MAX, u64::MAX],
+        ];
+
+        for vec in amounts.iter() {
+            assert!(compute_zero_balance(vec, vec, 2));
+        }
+
+        // Test when ours is empty
+        assert!(compute_zero_balance(&amounts[1], &amounts[1], 0));
+
+        // Test when theirs is empty
+        assert!(compute_zero_balance(
+            &amounts[1],
+            &amounts[1],
+            amounts[1].len()
+        ));
+
+        // Test random inputs and outputs
+        // Randomly distributed between ours and theirs allocation
+        for _ in 0..10 {
+            // Randomly generate number of amounts between 1 to 20
+            let length = rng.gen_range(1, 20);
+
+            // Randomly fill the amount vector
+            let mut amounts = vec![0; length];
+            for _ in 0..length {
+                // keep the amount value low for faster testing
+                amounts.push(rng.gen_range::<u64>(0, 10));
+            }
+
+            // Create an output amount vector such that
+            // input.sum() = output.sum(), but
+            // input.count() != output.count()
+            // We do this by generating an empty output vector
+            // of random size. Then randomly add 1 to random positions
+            // until the sum of the vector matches with the input sum
+            // This also creates 0 value allocations probabilistically
+
+            let input_sum: u64 = amounts.iter().sum();
+            let mut output_amounts = vec![0u64; rng.gen_range(1, 20)];
+            let output_length = output_amounts.len();
+            for _ in 0..input_sum {
+                output_amounts[rng.gen_range(0, output_length)] += 1u64;
+            }
```

Comment:
With the new value range this will not work efficiently (and will take a lot of time to compute), so I propose to generate `len-1` random numbers, set the last value of either input or output  (depending on which is larger, input or output) to their difference, and the last member of the other one - to zero

Reply:
This [comment](https://github.com/LNP-BP/rust-lnpbp/pull/68#discussion_r485060625). Yes there are many places having very large output vectors will be taxing. This will also create that many confidential amounts which will get huge in memory.

Reply:
We still need to change the value generation range, and adjusting with 1 would not work for matching the amounts. So this comment remains

Reply:
Not sure why it wont work. The value range that we have changed is reflected in the `input_sum`. It will try to push +1 until the input sum matches. This might take a long time. 
Trying to push random values and adjusting the last one might also result into negative value in the end as we don't have control over the random numbers. May be limiting the random numbers to say like 1000 for each value then the remaining into the last item? Something like that?

Reply:
In my first comment I had described an algo which will not result in negative problem

Reply:
What I meant is something like:
```rust
let diff = abs(sum(input) as i128 - sum(output) as i128);
if sum(input) < sum(output) {
   input[input.len()-1) = diff;
} else {
   output[output.len()-1] = diff;
}
```

Reply:
Ok got it.